hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellFlatMap.java

```diff
@@ -0,0 +1,468 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY CellIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.regionserver;
+
+import org.apache.hadoop.hbase.Cell;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Set;
+
+
+
+/**
+ * CellFlatMap stores a constant number of elements and is immutable after creation stage.
+ * Due to being immutable the CellFlatMap can be implemented as array.
+ * The actual array can be on- or off-heap and is implemented in concrete class derived from CellFlatMap.
+ * The CellFlatMap uses no synchronization primitives, it is assumed to be created by a
+ * single thread and then it can be read-only by multiple threads.
+ *
+ * The "flat" in the name, means that the memory layout of the Map is sequential array and thus
+ * requires less memory than ConcurrentSkipListMap.
+ */
+public abstract class CellFlatMap implements ConcurrentNavigableMap<Cell,Cell> {
+
+  private final Comparator<? super Cell> comparator;
+  private int minCellIdx   = 0;   // the index of the minimal cell (for sub-sets)
+  private int maxCellIdx   = 0;   // the index of the maximal cell (for sub-sets)
+  private boolean descending = false;
+
+  /* C-tor */
+  public CellFlatMap(Comparator<? super Cell> comparator, int min, int max, boolean d){
+    this.comparator = comparator;
+    this.minCellIdx = min;
+    this.maxCellIdx = max;
+    this.descending = d;
+  }
+
+  /* Used for abstract CellFlatMap creation, implemented by derived class */
+  protected abstract CellFlatMap createCellFlatMap(Comparator<? super Cell> comparator, int min,
+      int max, boolean descending);
+
+  /* Returns the i-th cell in the cell block */
+  protected abstract Cell getCellFromIndex(int i);
+
+  /**
+   * Binary search for a given key in between given boundaries of the array.
+   * Positive returned numbers mean the index.
+   * Negative returned numbers means the key not found.
+   * The absolute value of the output is the
+   * possible insert index for the searched key: (-1 * insertion point)
+   * @param needle The key to look for in all of the entries
+   * @return Same return value as Arrays.binarySearch.
+   */
+  private int find(Cell needle) {
+    int begin = minCellIdx;
+    int end = maxCellIdx - 1;
+
+    while (begin <= end) {
+      int mid = begin + ((end - begin) / 2);
+      Cell midCell = getCellFromIndex(mid);
+      int compareRes = comparator.compare(midCell, needle);
+
+      if (compareRes == 0) return mid;  // 0 means equals. We found the key
```

Comment:
Tried to fix it anywhere I saw it, other than that the code analyzer will draw my attention and I'll fix it there